README
NAME
    Regexp::Whitespace - Whitespace insensitivity for regexes

SYNOPSIS
        use Regexp::Whitespace ();

        my $re = Regexp::Whitespace->new( 'a b c', 'w' );

        "a   b\t\n c" =~ $re; # true

DESCRIPTION
        THIS IS A pre-alpha RELEASE.

    This module implements a /w modifier for regexes that provides
    whitespace insensitivity.

      qr/a b c/w    => qr/a\s+b\s+c/

  LIMITATIONS SO FAR

      * only supports a global 'w' modifier by now
      * only exact pieces are handled by now
        (which means things like / [ x] /xw are not transformed into
           / (?: \s+ | [x] ) /x

SEE ALSO
      perlre
      perlrecharclass (since Perl 5.10)

BUGS
    Please report bugs via CPAN RT
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Regexp-Whitespace or mailto:.

AUTHORS
    Adriano R. Ferreira, <ferreira@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2008 by Adriano R. Ferreira

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Regexp-Whitespace.

0.001_0  Mon Feb 21 2008
        - first release to CPAN
        - only simple things like qr/a b c/ works



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.2808_01
require 5.008;
use ExtUtils::MakeMaker;
WriteMakefile
(
          'DISTNAME' => 'Regexp-Whitespace',
          'VERSION_FROM' => 'lib/Regexp/Whitespace.pm',
          'PREREQ_PM' => {
                           'Attribute::Memoize' => '0',
                           'Clone' => '0.28',
                           'Scalar::Util' => '0',
                           'Test::More' => '0',
                           'YAPE::Regex' => '3.03',
                           'charnames' => '0',
                           'constant' => '0',
                           'if' => '0',
                           'ok' => '0',
                           'overload' => '0',
                           'strict' => '0',
                           'warnings' => '0'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [],
          'PL_FILES' => {}
        )
;


eg/extract_date.pl

use strict;
use warnings;

use Regexp::Whitespace ();

# that's a simple example to extract a piece of text from
# within text samples with arbitrary spaces
# (coming from example from stripping tags of HTML files)
#
# The date position is marked around a "fixed" text
# like "date of event which ..." but spaces should be
# replaced by \s+
#
# With Regexp::Whitespace, instead of writing
# "date \s+ of \s+ event \s+ ...", the regex may look
# like simple text and the \w modifier does the magic.

my $re = q/(?imsx:
             \G \s*
                    ( .*?                        # advance until you find
                      ((?-x)date of event which requires filing of (?x: this | the )  statement)
                      .*?
                      $                          # and to the end of line
                    )
            )/;

my $qr = Regexp::Whitespace->new( $re, 'w' );

my @samples = (

q{
                                 August 31, 2007
             (Date of event which requires filing of this statement)
},

q{
                  January
                    14, 2008




                  Date
                    of Event Which Requires Filing of the
                    Statement
}

);

for my $text ( @samples ) {
    if ( $text =~ $qr ) {
        my ($piece, $literal) = ($1, $2);
        $piece =~ s/ \Q$literal\E //msx; # remove the fixed part
        $piece =~ s/ [^\w.]+ \z //msx; # remove "weird" trailing chars (like '(') 
        $piece =~ s/ \s+ / /msgx; # normalize spaces

        print "matched, date found: $piece\n";

    }
}


lib/Regexp/Whitespace.pm

package Regexp::Whitespace;

use 5.008;
use strict;
use warnings;

our $VERSION = '0.001_0';
our @ISA;

BEGIN {
  @ISA = qw( Regexp );
}

use Regexp::Whitespace::Builder ();

use Attribute::Memoize;
use Scalar::Util qw( refaddr );

use overload 
    '""'     => \&to_s,
    fallback => 1;

sub new {
    my $self = shift;
    my $proto = ref $self || $self;
    my $qr = $self->_factory->build(@_);
    my $s = "$qr";
    bless $qr, $proto;
    $qr->raw( qq{/$_[0]/w} ); # FIXME
    $qr->_re_string( $s ); # FIXME
    return $qr;
}

sub _factory :Memoize {
    return 'Regexp::Whitespace::Builder';
}

{
    # FIXME an incomplete inside-out approach
    my %raw_of;
    my %string_of; # I am using this because I could not
                   # invoke the default Regexp stringification
                   # after blessing
    sub raw {
        my $self = shift;
        if ( @_ ) {
            $raw_of{refaddr $self} = shift;
        }
        return $raw_of{refaddr $self};
    }
    # the stringfication of the underlying regex
    sub _re_string {
        my $self = shift;
        if ( @_ ) {
            $string_of{refaddr $self} = shift;
        }
        return $string_of{refaddr $self};
    }
    # FIXME: needs DESTROY
}

sub to_s {
    return shift->raw;
}


#sub import {
#    use overload;
#
#    overload::constant
#      qr => sub {
#          my ($raw) = @_;
#          ...
#      };
#}

1;


lib/Regexp/Whitespace/Parser.pm

use strict;
use warnings;

use Clone;

package Regexp::Whitespace::Parser;

our $VERSION = '0.001_0';

use YAPE::Regex qw( Regexp::Whitespace::Parser );

package Regexp::Whitespace::Parser::Element;

our @ISA;

BEGIN {
  push @ISA, qw( Clone );
}

sub convert {
    # clone by default
    return shift->clone;
}

package Regexp::Whitespace::Parser::container;

sub convert {
    my $self = shift;
    my $clone = $self->clone(1); # shallow copy
    my @content = map { $_->convert } @{$self->{CONTENT}};
    $clone->{CONTENT} = \@content;
    return $clone;
}

# register this package into @ISA of container types
{
    my @container_types = qw( cut lookahead lookbehind group capture );
    no strict 'refs';
    for my $type (@container_types) {
        unshift @{ 'Regexp::Whitespace::Parser::' . $type . '::ISA' }, __PACKAGE__;
    }
}

package Regexp::Whitespace::Parser::exact;

sub convert {
    my $self = shift;
    my $exact = $self->exact_text;

    # are there ocurrences of \s ?
    if ( $exact =~ /\s/ ) {
        my @pieces;

        if ( length $exact > 1 ) {
            # note: only 'text' types need this loop,
            #  being the only one whose exact text may have
            #  a length greater than 1

            # assertions
            die "panic: quantity modifier should not present for multi-character text" if $self->quant;
            die "panic: non-greedy modifier should not present for multi-character text" if $self->ngreed;

            LOOP : {
                if ( $exact =~ / \G \z /xgc ) {
                    last LOOP;
                }
                if ( $exact =~ / \G \s+ /xgc ) {
                    # replace matches of /\s+/ with a macro '\s+'
                    push @pieces, Regexp::Whitespace::Parser::macro->new( 's', '+', '' );
                    redo LOOP;
	        }
                if ( $exact =~ / \G (\S+) /xgc ) {
	            push @pieces, Regexp::Whitespace::Parser::text->new( $1, '', '' );
                    redo LOOP;
                }
            }
        } else {
           # FIXME: these conversion rules needs checking
           #   s is any char that matches /\s/
           # s      becomes   \s+
           # s?     becomes   \s*
           # s*     becomes   \s*
           # s{0}   becomes   \s*
           # s{0,N} becomes   \s*
           # s{M,N} becomes   \s+
           # s+     becomes   \s+
           # 
           # the non-greedy flag is kept (don't know if that's correct)
           my ($q, $ng) = ($self->quant, $self->ngreed);
           my $nq = ( $q =~ /\A ( [?*] | [{]0 ) /x ) ? '*' : '+';
           return Regexp::Whitespace::Parser::macro->new( 's', $nq, $ng );

           # TODO: some tests would be nice
        }

        return @pieces;

    } else {
        # no needed conversion
        return $self->clone;
    }
}

# register this package into @ISA of exact types
{
    my @exact_types = qw( text oct hex slash ctrl named ); # ?! utf8hex
    no strict 'refs';
    for my $type (@exact_types) {
        unshift @{ 'Regexp::Whitespace::Parser::' . $type . '::ISA' }, __PACKAGE__;
    }
}

package Regexp::Whitespace::Parser::text;

sub exact_text {
    return shift->{TEXT};
}

package Regexp::Whitespace::Parser::oct;

sub exact_text {
    return chr oct(shift->{TEXT});
}

package Regexp::Whitespace::Parser::hex;

sub exact_text {
    return chr hex(shift->{TEXT});
}

#package ...::utf8hext; # FIXME wait for new release of YAPE::Regex

package Regexp::Whitespace::Parser::slash;

my %known_sequences = (
   't' => "\t",
   'n' => "\n",
   'r' => "\r",
   'a' => "\a",
   'f' => "\f",
   'b' => "\b",
   'e' => "\e",
);

sub exact_text {
    my $t = shift->{TEXT};
    return $known_sequences{$t} || $t;
}

package Regexp::Whitespace::Parser::ctrl;

sub exact_text {
    my $t = shift->{TEXT};
    return chr( ord(uc $t) ^ 0x40 );
}

package Regexp::Whitespace::Parser::named;

sub exact_text {
    require charnames;
    return charnames::vianame(shift->{TEXT});
}


1;



lib/Regexp/Whitespace/Builder.pm

package Regexp::Whitespace::Builder;

use 5.008;
use strict;
use warnings;

our $VERSION = '0.001_0';

use Regexp::Whitespace::Parser ();

use constant _debug => 0;

use if _debug, DDS => 'Dump';
#use charnames ':full'; # FIXME make it conditional

sub build {
    my $self = shift;
    my $re = shift;
    my $flags = shift;
    die "flags must be 'w' in this early version" if $flags ne 'w';

    my $p = Regexp::Whitespace::Parser->new( $re );
    $p->parse;
    my $t = $p->top;
    die "could not parse $re: " . $p->error unless $t || $p->error;

    Dump( $t ) if _debug;

    # walk the RE tree doing transformations [\s] -> \s+
    my $nt = $t->convert;

    Dump( $nt ) if _debug;

    my $regex = $nt->fullstring;

    # FIXME: there's a bug here if we're using \N{named} escapes
    #   because they are replaced at compile time and
    #   would need something like eval qq{use charnames ':full'; qr/$s/}
    #
    # Instead, we should mark (TODO) there is named escapes involved
    #   and do a substitution   s/\N{([^}])}/charnames::vianame($1)/ge
    #   before mounting the regex

    if ( $regex =~ /\A [(][?]-imsx: (.*) [)] \z/sx ) { # FIXME
       #warn "# GOT";
       return qr/$1/;
    }
    #warn "# FALLBACK";
    return qr/$regex/;
}

1;


Build.PL

use Module::Build;

Module::Build->new(
  dist_name => 'Regexp-Whitespace',
  dist_version_from => 'lib/Regexp/Whitespace.pm',
  license => 'perl',
  author => 'Adriano Ferreira <ferreira@cpan.org>',
  requires => {
    'perl'                    => 5.008,
    'strict'                  => 0,
    'warnings'                => 0,
    'constant'                => 0,
    'if'                      => 0,

    'overload'                => 0,

    'Attribute::Memoize'      => 0,
    'Scalar::Util'            => 0,
    'Clone'                   => 0.28,
    'YAPE::Regex'             => 3.03,
    'charnames'               => 0,

  },
  build_requires => {
    'Test::More'              => 0,
    'ok'                      => 0, # from Test-use-ok
  },

  create_makefile_pl => 'traditional',
  create_readme => 1,

)->create_build_script;
